### ISSUE TITLE
Lazy log message formatting

### ISSUE THREAD

**Issue by TheJJ** [SFTtech/openage#1001]
Our logging system does not support lazy message formatting, like the python `logging` module.

``` python
from log import info

# current version: always performs string formatting
var = "with formatting"
info("some message %s" % var)

# new version: should only format the string if the message is actually emitted
var = "lazyness"
info("everything is better™ with ", var)
```

The functions are in `openage/log/__init__.py` and then relay the messages to `libopenage`.
The formatting of messages should thus happen (either in C++ or the python part, not sure) **only if** the log level of the message is higher than the minimum log level of a sink that gets the message. 

This is an optimization, so no real priority :wink:

**Comment by kev946** [comment]
Are we just trying to do lazy formatting on the messages sent from the Python side?

**Comment by TheJJ** [comment]
We can do both, in C++ and Python. The C++ `operator <<` could be abused for that, and we would have to store references to the message parts, to merge them later if the message has to be formatted.

We could add that C++ (which controls the loglevel) notifies Python of the current level, so messages are not even transported C++ if the message is not important enough.

**Comment by TheJJ** [comment]
I added a possible approach section above, with the main idea to create a backend for our C++ logger for the python `logging` module.

**Comment by kev946** [comment]
For the Python side, maybe we could create a `CppHandler` class derived from [`Handler`](https://docs.python.org/3/library/logging.html#logging.Handler). We override [`setLevel`](https://docs.python.org/3/library/logging.html#logging.Handler.setLevel) and [`emit`](https://docs.python.org/3/library/logging.html#logging.Handler.emit) to call into the C++ side. Does this sound good?

**Comment by TheJJ** [comment]
Yes, except that we have to call `setLevel` from the C++ side so the C++-level is set in the python world.
